src/vs/workbench/contrib/chat/browser/chatContentParts/toolInvocationParts/chatTerminalToolProgressPart.ts

```diff
@@ -106,42 +142,325 @@ export class ChatTerminalToolProgressPart extends BaseChatToolInvocationSubPart
+			if (resolvedCommand?.endMarker) {
+				this._ensureShowOutputAction();
+			}
+			return resolvedCommand;
+		};
+
+		const attachCommandDetection = (commandDetection: ICommandDetectionCapability | undefined) => {
+			commandDetectionListener.clear();
+			if (!commandDetection) {
+				return;
+			}
+
+			const resolvedImmediately = tryResolveCommand();
+			if (resolvedImmediately?.endMarker) {
+				return;
+			}
+
+			commandDetectionListener.value = commandDetection.onCommandFinished(() => {
+				this._ensureShowOutputAction();
+				commandDetectionListener.clear();
+			});
+		};
+
+		attachCommandDetection(terminalInstance.capabilities.get(TerminalCapability.CommandDetection));
+		this._register(terminalInstance.capabilities.onDidAddCommandDetectionCapability(cd => attachCommandDetection(cd)));
+
 		const instanceListener = this._register(terminalInstance.onDisposed(() => {
+			if (this._terminalInstance === terminalInstance) {
+				this._terminalInstance = undefined;
+			}
+			commandDetectionListener.clear();
 			this._actionBar.clear();
-			instanceListener?.dispose();
+			this._focusAction.clear();
+			const keepOutputAction = !!this._terminalData.output?.html;
+			this._showOutputActionAdded = false;
+			if (!keepOutputAction) {
+				this._showOutputAction.clear();
+			}
+			this._ensureShowOutputAction();
+			instanceListener.dispose();
 		}));
 	}
+
+	private async _toggleOutput(expanded: boolean): Promise<boolean> {
+		const currentlyExpanded = this._outputContainer.classList.contains('expanded');
+		if (expanded === currentlyExpanded) {
+			this._showOutputAction.value?.syncPresentation(currentlyExpanded);
+			return false;
+		}
+
+		this._setOutputExpanded(expanded);
+
+		if (!expanded) {
+			this._layoutOutput();
+			this._showOutputAction.value?.syncPresentation(false);
+			return true;
+		}
+
+		const didCreate = await this._renderOutputIfNeeded();
+		this._layoutOutput();
+		this._scrollOutputToBottom();
+		if (didCreate) {
+			this._scheduleOutputRelayout();
+		}
+		this._showOutputAction.value?.syncPresentation(expanded);
+		return true;
+	}
+
+	private _setOutputExpanded(expanded: boolean): void {
+		this._outputContainer.classList.toggle('expanded', expanded);
+		this._outputContainer.classList.toggle('collapsed', !expanded);
+		this._titlePart.classList.toggle('expanded', expanded);
+	}
+
+	private async _renderOutputIfNeeded(): Promise<boolean> {
+		if (this._outputContent) {
+			this._ensureOutputResizeObserver();
+			return false;
+		}
+
+		if (!this._terminalInstance) {
+			const resource = this._getTerminalResource();
+			if (resource) {
+				this._terminalInstance = this._terminalService.getInstanceFromResource(resource);
+			}
+		}
+		const output = await this._collectOutput(this._terminalInstance);
+		const content = this._renderOutput(output);
+		const theme = this._terminalInstance?.xterm?.getXtermTheme();
+		if (theme) {
+			const inlineTerminal = content.querySelector('div');
```

Comment:
Should use the `h` helper, and keep track of elements we care about. We should very rarely query sub-elements like this. Doing this should also make the `if (inlineTerminal)` on the following line unnecessary.